### PR TITLE
add option for preloader for xrdcp stageout

### PR DESCRIPTION
**Pull request by hufnagel**
For stageout from certain HPC we want to have a preloader in front of every call to a TCP application to tunnel the traffic. Allow to configure this in SITECONF.

#### Status
ready

#### Description
Adds an option for a preloader in front of every xrdcp and xrdfs call that can be specified in the site-local-conf stageout section.

#### Is it backward compatible (if not, which system it affects?)
YES (only sites that set the new option will see any effect)

#### Related PRs
NONE

#### External dependencies / deployment changes
NONE

### CODE REVIEW

File: src/python/WMCore/Storage/Backends/XRDCPImpl.py [hunk 1]
```diff
@@ -133,18 +138,19 @@ def createStageOutCommand(self, sourcePFN, targetPFN, options=None, checksums=No
 
         (_, host, path, _) = self.splitPFN(remotePFN)
 
+        removeCommand = ""
```
Comment (amaltaro) [inline src/python/WMCore/Storage/Backends/XRDCPImpl.py:141]: This line seems to not be necessary.

File: src/python/WMCore/Storage/Backends/XRDCPImpl.py [hunk 2]
```diff
@@ -115,7 +117,10 @@ def createStageOutCommand(self, sourcePFN, targetPFN, options=None, checksums=No
         if args.wma_disablewriterecovery:
             copyCommand += "env XRD_WRITERECOVERY=0 "
 
-        copyCommand += "%s --force --nopbar " % xrdcpExec
+        if args.wma_preload:
+            self.preload = "%s " % args.wma_preload
+
+        copyCommand += "%s%s --force --nopbar " % (self.preload, xrdcpExec)
```
Comment (amaltaro) [inline src/python/WMCore/Storage/Backends/XRDCPImpl.py:123]: I think all these commands would look cleaner if we add an extra space in every single command, instead of stucking two format strings altogether. So, something like:
```
copyCommand += "%s %s --force --nopbar " % (self.preload, xrdcpExec)
```
and the same to all the other changes below.

That also means there is no need for the extra space above in L121.

**Review by amaltaro — COMMENTED** [review]
Thanks for providing this functionality, Dirk.

Code looks good to me, however, if you also agree with me that a command like (or source code):
```
"%s xrdfs %s rm %s"
```

goes much better with the eyes/brain, than
```
"%sxrdfs %s rm %s"
```

then please update this code.

File: src/python/WMCore/Storage/Backends/XRDCPImpl.py (hunk 2 above)
Comment (hufnagel) [reply src/python/WMCore/Storage/Backends/XRDCPImpl.py:123]: If you add the extra space you break the unit tests since the "no preloader set" option now looks different from before (it has the empty space).

Comment (hufnagel) [reply src/python/WMCore/Storage/Backends/XRDCPImpl.py:123]: Only other option I see is to have a conditional for every line that includes the preloader and make one version with preloader and one version without.

Comment (hufnagel) [reply src/python/WMCore/Storage/Backends/XRDCPImpl.py:123]: Or we merge ignoring the unit test failures since after the merge the unit tests should work again (the version with the empty space becomes the new default).

Comment (amaltaro) [reply src/python/WMCore/Storage/Backends/XRDCPImpl.py:123]: Perhaps we could return the stage out command like:
```
return copyCommand.strip()
```
? That should resolve the unit test failures.

**Comment by hufnagel** [comment]
Unit tests pass and I think this looks cleaner than before.

File: src/python/WMCore/Storage/Backends/XRDCPImpl.py [hunk 3]
```diff
@@ -115,7 +117,14 @@ def createStageOutCommand(self, sourcePFN, targetPFN, options=None, checksums=No
         if args.wma_disablewriterecovery:
             copyCommand += "env XRD_WRITERECOVERY=0 "
 
-        copyCommand += "%s --force --nopbar " % xrdcpExec
+        if args.wma_preload:
+            xrdcpCmd = "%s %s" % (args.wma_preload, xrdcpExec)
+            self.xrdfsCmd = "%s xrdfs" % args.wma_preload
+        else:
+            xrdcpCmd = xrdcpExec
+            self.xrdfsCmd = "xrdfs"
```
Comment (amaltaro) [inline src/python/WMCore/Storage/Backends/XRDCPImpl.py:125]: This line isn't really needed.

**Review by amaltaro — COMMENTED** [review]
Thanks Dirk. Code looks better indeed. I left only one more comment for improvement.

I'm applying this patch - as is - to the hepcloud agent (cmssrv217) just so we can speed up the testing process.

File: src/python/WMCore/Storage/Backends/XRDCPImpl.py (hunk 3 above)
Comment (hufnagel) [reply src/python/WMCore/Storage/Backends/XRDCPImpl.py:125]: It's redundant, yes. Makes the code a bit easier to read and understand IMO.

Comment (amaltaro) [reply src/python/WMCore/Storage/Backends/XRDCPImpl.py:125]: Still it's an unneeded code duplication. Anyways, not a big deal.

**Comment by amaltaro** [comment]
I remember you saying it works fine, even though there was a unrelated issue at those resources. So I'm merging it.
Please let me know if any follow up is required.
